containers/src/Data/IntMap/Internal.hs

```diff
@@ -62,2904 +275,1189 @@
-    go Nil = not_found
-
-    not_found = error ("IntMap.!: key " ++ show k ++ " is not an element of the map")
-
--- | /O(min(n,W))/. The expression @('findWithDefault' def k map)@
+    start (IntMap Empty) = Nothing
+    start (IntMap (NonEmpty min minV node))
+        | outOfMinBound k min = Nothing
+        | k == boundKey min = Just minV
+        | otherwise = goL (xor k min) node
+
+    goL !_ Tip = Nothing
+    goL !xorCache (Bin max maxV l r)
+        | inMaxBound k max = if xorCache < xorCacheMax
+                    then goL xorCache l
+                    else goR xorCacheMax r
+        | outOfMaxBound k max = Nothing
+        | otherwise = Just maxV
+      where xorCacheMax = xor k max
+
+    goR !_ Tip = Nothing
+    goR !xorCache (Bin min minV l r)
+        | inMinBound k min = if xorCache < xorCacheMin
+                    then goR xorCache r
+                    else goL xorCacheMin l
+        | outOfMinBound k min = Nothing
+        | otherwise = Just minV
+      where xorCacheMin = xor k min
+
+-- | /O(min(n,W))/. The expression @'findWithDefault' def k map@
 -- returns the value at key @k@ or returns @def@ when the key is not an
 -- element of the map.
 --
 -- > findWithDefault 'x' 1 (fromList [(5,'a'), (3,'b')]) == 'x'
 -- > findWithDefault 'x' 5 (fromList [(5,'a'), (3,'b')]) == 'a'
-
--- See Note: Local 'go' functions and capturing]
 findWithDefault :: a -> Key -> IntMap a -> a
-findWithDefault def !k = go
+findWithDefault def k = k `seq` start
   where
-    go (Bin p m l r) | nomatch k p m = def
-                     | zero k m  = go l
-                     | otherwise = go r
-    go (Tip kx x) | k == kx   = x
-                  | otherwise = def
-    go Nil = def
+    start (IntMap Empty) = def
+    start (IntMap (NonEmpty min minV node))
+        | outOfMinBound k min = def
+        | k == boundKey min = minV
+        | otherwise = goL (xor k min) node
+
+    goL !_ Tip = def
+    goL !xorCache (Bin max maxV l r)
+        | inMaxBound k max = if xorCache < xorCacheMax
+                    then goL xorCache l
+                    else goR xorCacheMax r
+        | outOfMaxBound k max = def
+        | otherwise = maxV
+      where xorCacheMax = xor k max
+
+    goR !_ Tip = def
+    goR !xorCache (Bin min minV l r)
+        | inMinBound k min = if xorCache < xorCacheMin
+                    then goR xorCache r
+                    else goL xorCacheMin l
+        | outOfMinBound k min = def
+        | otherwise = minV
+      where  xorCacheMin = xor k min
 
 -- | /O(log n)/. Find largest key smaller than the given one and return the
 -- corresponding (key, value) pair.
 --
 -- > lookupLT 3 (fromList [(3,'a'), (5,'b')]) == Nothing
 -- > lookupLT 4 (fromList [(3,'a'), (5,'b')]) == Just (3, 'a')
-
--- See Note: Local 'go' functions and capturing.
 lookupLT :: Key -> IntMap a -> Maybe (Key, a)
-lookupLT !k t = case t of
-    Bin _ m l r | m < 0 -> if k >= 0 then go r l else go Nil r
-    _ -> go Nil t
+lookupLT k = k `seq` start
```

Comment:
Since GHC doesn't do worker-wrapper on sums, do it manually. Don't use my wretched names.

```haskell
data MKV a = MKV !Key a | NoMKV
lookupLT !k im0 = mkvToMaybe $ start im0
```

Reply:
Er, sorry, that last comment isn't the way to do it. Use an *unboxed* sum where available as the intermediate result.

Reply:
My comment keeps getting lost... Do it with unboxed sums and products.

Reply:
Sorry, I've never used GHC's unboxed sum support, and I don't know how to apply it for this project. In particular, to handle pre-8.2, should I have two different versions of a function depending on whether unboxed sums are supported? I didn't find any other code in `containers` using them to draw inspiration from.

(I also don't think specifically `lookupLT` and friends need them; the new representation has `O(1)` min/max access, so we choose between `Nothing` and `Just` before calling anything recursive, which GHC should handle just fine. However, there are definitely other functions that would benefit from unboxed sums: this is what the old `SketchyIntMap` hack was, essentially.)

Reply:
@treeowl How about leaving optimizations for less important functions for future work? Optimization tasks like this one could be a good way to get other/new contributors involved with the new code.

In fact, @gereeter, could you add two TODO-lists to the PR description: One for things that need to be done before the merge, and one for future work?

Reply:
Resolving as future work in #698.